client/my-sites/themes/helpers.js

```diff
@@ -112,3 +114,54 @@ export function getSubjectsFromTermTable( filterToTermTable ) {
 			return obj;
 		}, {} );
 }
+
+/**
+ * Interlace WP.com themes with WP.org themes according to the logic below:
+ * - WP.org themes are only included if there is a search term.
+ * - If the search term has an exact match (either a WP.com or a WP.org theme), that theme is the first result.
+ * - WP.com themes are prioritized over WP.org themes.
+ * - Retired WP.org themes or duplicate WP.org themes (those that are also WP.com themes) are excluded.
+ *
+ * @param wpComThemes List of WP.com themes.
+ * @param wpOrgThemes List of WP.org themes.
+ * @param searchTerm Search term.
+ * @param isLastPage Whether the list of WP.com themes has reached the last page.
+ */
+export function interlaceThemes( wpComThemes, wpOrgThemes, searchTerm, isLastPage ) {
+	const isMatchingTheme = ( theme ) => {
+		if ( ! searchTerm ) {
+			return false;
+		}
+		return (
+			theme.name?.toLowerCase() === searchTerm?.toLowerCase() ||
+			theme.id?.toLowerCase() === searchTerm?.toLowerCase()
+		);
+	};
+
+	const includeWpOrgThemes = !! searchTerm;
+	const wpComThemesSlugs = wpComThemes.map( ( theme ) => theme.id );
+	const validWpOrgThemes = includeWpOrgThemes
+		? wpOrgThemes.filter(
+				( theme ) =>
+					! wpComThemesSlugs.includes( theme?.id?.toLowerCase() ) && // Avoid duplicate themes. Some free themes are available in both wpcom and wporg.
+					! RETIRED_THEME_SLUGS_SET.has( theme?.id?.toLowerCase() ) // Avoid retired themes.
+		  )
+		: [];
+
+	const matchingTheme = wpComThemes.find( isMatchingTheme );
+	const restWpComThemes = matchingTheme
+		? wpComThemes.filter( ( theme ) => theme.id !== matchingTheme.id )
+		: wpComThemes;
+	const matchingWpOrgTheme = validWpOrgThemes.find( isMatchingTheme );
+	const restWpOrgThemes = matchingWpOrgTheme
+		? validWpOrgThemes.filter( ( theme ) => theme.id !== matchingWpOrgTheme.id )
+		: validWpOrgThemes;
+
+	return [
+		...( matchingTheme ? [ matchingTheme ] : [] ),
+		...( matchingWpOrgTheme ? [ matchingWpOrgTheme ] : [] ),
+		...restWpComThemes,
+		// Include WP.org themes after the last page of WP.com themes.
+		...( isEnabled( 'themes/interlaced-dotorg-themes' ) && isLastPage ? restWpOrgThemes : [] ),
```

Comment:
Should we also make the search results for exact coincidence use this config value?

Reply:
Nope, that would break the existing behavior in production which displays WP.org results when there is an exact match.